cms/djangoapps/contentstore/models.py

```diff
@@ -327,6 +371,36 @@ def filter_links(
         )
         if ready_to_sync is not None:
             result = result.filter(ready_to_sync=ready_to_sync)
+
+        if use_top_level_parents is not None:
+            # Get objects without top_level_parent_usage_key
+            non_top_level_objects = result.filter(top_level_parent_usage_key=UsageKeyField.Empty)
+            non_top_level_ids = non_top_level_objects.values_list('id', flat=True)
+
+            # Get the non-null top_level_parent_usage_key
+            # Note: using `exclude()` raise `TypeError`
+            top_level_keys = result.exclude(id__in=non_top_level_ids).values_list(
+                'top_level_parent_usage_key', flat=True
+            ).distinct()
+            top_level_objects = cls.objects.filter(
+                downstream_usage_key__in=top_level_keys,
+            ).select_related(*RELATED_FIELDS)
+
+            final_objects = list(chain(top_level_objects, non_top_level_objects))
+            final_ids = [obj.id for obj in final_objects]
+
+            result = cls.objects.filter(id__in=final_ids).select_related(*RELATED_FIELDS).annotate(
+                ready_to_sync=(
+                    GreaterThan(
+                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
+                        Coalesce("version_synced", 0)
+                    ) & GreaterThan(
+                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
+                        Coalesce("version_declined", 0)
+                    )
+                )
+            )
```

Comment:
With help of some AI and django ORM, we can simplify this as below. The same can be done with `ComponentLinks.filter_links` function.

```suggestion
        # Handle top-level parents logic
        if use_top_level_parents is not None:
            # Subquery to find top-level parent usage keys
            top_level_parent_subquery = cls.objects.filter(
                downstream_usage_key=OuterRef('top_level_parent_usage_key')
            ).values('id')

            # Annotate with a flag indicating if the object has a top-level parent
            result = result.annotate(
                has_top_level_parent=Exists(top_level_parent_subquery)
            )

            # If use_top_level_parents is True, filter to get only top-level parents or objects without parents
            return result.filter(
                Q(has_top_level_parent=True) | Q(top_level_parent_usage_key=UsageKeyField.Empty)
            )
```

Reply:
This code does not replace descendants with their top-level parents. See https://github.com/openedx/edx-platform/pull/37076#discussion_r2240671472

I have tried using the `top_level_parent_subquery` in my code, but it has become complicated. Do you think the code needs major changes?

Reply:
Ahh, that is good then. We could however reduce a call to database by annotating `top_level_objects` query with `ready_to_sync` and skip the final result call, something like below **(Note: you will have to shuffle this [list](https://github.com/open-craft/edx-platform/blob/806cb4f3f1260905b3306d03dbf86dd85786dc82/cms/djangoapps/contentstore/rest_api/v2/views/tests/test_downstreams.py#L1125) in test with the last one to pass.)**

```python
            top_level_objects = cls.objects.filter(
                downstream_usage_key__in=top_level_keys,
            ).select_related(*RELATED_FIELDS).annotate(
                ready_to_sync=(
                    GreaterThan(
                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
                        Coalesce("version_synced", 0)
                    ) & GreaterThan(
                        Coalesce("upstream_container__publishable_entity__published__version__version_num", 0),
                        Coalesce("version_declined", 0)
                    )
                )
            )

            result = list(chain(top_level_objects, non_top_level_objects))
```

And it would be nice to create a function that annotates the query with `ready_to_sync` field as it seems to be repeated here.

Reply:
Thanks! Updated: https://github.com/openedx/edx-platform/pull/37076/commits/966c998b2d24d564cd1d17eb3cb238034979d809